### PR TITLE
Shorten virtual dispatch tables

### PR DESCRIPTION
**Pull request by smowton**
This shortens the dispatch tables produced by remove_virtual_functions by eliminating duplicate dispatch entries, and by eliminating calls that don't have any possible targets that we're aware of.

Needs rebasing after the factor-get-class-identifier PR is merged. Meanwhile, the real diff is: https://github.com/smowton/cbmc/compare/factor_class_identifier...smowton:shorten_virtual_dispatch_tables?expand=1

### CODE REVIEW

File: src/goto-programs/goto_inline.cpp [hunk 1]
```diff
@@ -826,8 +826,17 @@ void goto_partial_inline(
         it=goto_functions.function_map.begin();
         it!=goto_functions.function_map.end();
         it++)
-      if(it->second.body_available())
+    {
```
Comment (peterschrammel) [inline src/goto-programs/goto_inline.cpp]: This seems to be an unrelated change.

File: src/goto-programs/remove_virtual_functions.cpp [hunk 2]
```diff
@@ -49,9 +50,9 @@ class remove_virtual_functionst
 
   typedef std::vector<functiont> functionst;
   void get_functions(const exprt &, functionst &);
+  void get_child_functions_rec(const irep_idt &, const symbol_exprt &,
```
Comment (peterschrammel) [inline src/goto-programs/remove_virtual_functions.cpp]: cpplint

File: src/goto-programs/remove_virtual_functions.cpp [hunk 3]
```diff
@@ -227,6 +211,35 @@ void remove_virtual_functionst::remove_virtual_function(
...
+  functionst &functions)
+{
+  auto findit=class_hierarchy.class_map.find(this_id);
+  if(findit==class_hierarchy.class_map.end())
+    return;
+
+  for(const auto & child : findit->second.children)
+  {
+    exprt method=get_method(child, component_name);
+    functiont function;
```
Comment (peterschrammel) [inline src/goto-programs/remove_virtual_functions.cpp]: to avoid copying, push_back(functiont()) first and then get a reference to .back()

File: src/goto-programs/goto_inline.cpp (hunk 1 above)
Comment (smowton) [inline src/goto-programs/goto_inline.cpp]: This isn't in the PR. Please click through to the real diff to see what's actually in it.

File: src/goto-programs/remove_virtual_functions.cpp (hunk 2 above)
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Not in PR

File: src/goto-programs/remove_virtual_functions.cpp (hunk 3 above)
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Not in PR

**Comment by smowton** [comment]
The last one of these I believe is part of one of the other PRs though; I will apply it there.

**Comment by smowton** [comment]
Rebased. Expecting Travis failure due to https://github.com/diffblue/cbmc/pull/483

File: regression/cbmc-java/virtual7/test.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (peterschrammel) [inline regression/cbmc-java/virtual7/test.java]: Is it necessary to have so many blank lines in this file?

File: src/goto-programs/remove_virtual_functions.cpp [hunk 5]
```diff
@@ -142,46 +142,54 @@ void remove_virtual_functionst::remove_virtual_function(
   // If necessary, cast to the last candidate function to
   // get the object's clsid. By the structure of get_functions,
   // this is the parent of all other classes under consideration.
-  symbol_typet suggested_type(functions.back().class_id);
+  auto base_classid=functions.back().class_id;
```
Comment (peterschrammel) [inline src/goto-programs/remove_virtual_functions.cpp]: is it necessary to copy here?

**Comment by smowton** [comment]
Comments addessed

File: src/goto-programs/remove_virtual_functions.cpp [hunk 6]
```diff
@@ -142,46 +144,54 @@ void remove_virtual_functionst::remove_virtual_function(
...
+      }
+      else
+      {
+        // No definition for this type; shouldn't be possible...
+        t1->make_assertion(false_exprt());
+      }
+      insertit.first->second=t1;
+      // goto final
+      goto_programt::targett t3=new_code_calls.add_instruction();
+      t3->make_goto(t_final, true_exprt());
```
Comment (tautschnig) [inline src/goto-programs/remove_virtual_functions.cpp]: All those additional instructions should have their `source_location` set, unless that's for some reason not possible? (Not a Java expert here.)

**Comment by smowton** [comment]
Done

**Comment by tautschnig** [comment]
@smowton Thanks! @peterschrammel are you happy with the changes?

**Comment by peterschrammel** [comment]
@tautschnig, yes. I'm waiting for travis to give the final go.

**Comment by smowton** [comment]
@peterschrammel do you have the power to reset these broken Travis jobs?

**Comment by tautschnig** [comment]
@smowton Isn't the "virtual7" regression test genuinely broken? I.e., not a matter of resetting, but rather in need of fixing in this pull request?

**Comment by smowton** [comment]
When I wrote this the job was crashed. Now virtual7 fails, because of different set ordering between platforms. I think the desired match "if E then goto X; if B then goto X; if D then goto Y; if C then goto Y;" can't be expressed in our test descriptor format.

**Comment by smowton** [comment]
Pushed an approximation to the desired test. Waiting for Travis again.

**Comment by peterschrammel** [comment]
Your condition looks regular, though.

**Comment by smowton** [comment]
Not sure what that means?

The situation was, on some platforms it would say B -> 1, E -> 1, C -> 2, D -> 2, and on others it would switch 1 and 2 around (doesn't make any difference). I can't write a regex that matches up the GOTOs per line, so I just ask for {B, C, D, E} -> [12] and no "GOTO 4" as loose evidence that the desired smaller CFG has been constructed.

**Comment by peterschrammel** [comment]
I remember @thk123 having an extension of test.pl for multiline matching...

**Comment by tautschnig** [comment]
... which doesn't currently work on Mac OS. See #358.

**Comment by smowton** [comment]
I think regardless I can't specify two things matching, except by exhaustively listing the possibilities?

**Comment by thk123** [comment]
@smowton Ah yup sorry didn't actually check what you wanted to do - just saw it was mentioned :P This I think would be another case where the fact that all the tests are using the same script is limiting the expressiveness of our tests.

**Comment by smowton** [comment]
@peterschrammel / @kroening are you happy to merge as-is?
